File: backend/app/services/contract_service.py
from sqlalchemy.orm import Session
from typing import List, Optional
from app.models import Contract, Department, Vendor
from app.schemas.contracts import ContractSummary, ContractDetail, RiskOut, RiskFlagOut
from ml.risk_engine.engine import RiskEngine
from app.core.config import settings


class ContractService:
    def __init__(self, db: Session):
        self.db = db
        self.risk_engine = RiskEngine()

    def get_contracts(
        self,
        department_id: Optional[int] = None,
        vendor_id: Optional[int] = None,
        limit: int = 50,
        offset: int = 0
    ) -> List[ContractSummary]:
        """Get list of contracts with optional filtering."""
        query = self.db.query(Contract)

        if department_id:
            query = query.filter(Contract.department_id == department_id)
        if vendor_id:
            query = query.filter(Contract.vendor_id == vendor_id)

        contracts = query.offset(offset).limit(limit).all()

        return [
            ContractSummary(
                id=c.id,
                contract_number=c.contract_number,
                title=c.title,
                contract_date=c.contract_date,
                department_id=c.department_id,
                vendor_id=c.vendor_id,
                estimate_value=c.estimate_value,
                award_value=c.award_value,
                risk_level=self._get_risk_level(c)
            )
            for c in contracts
        ]

    def get_contract(self, contract_id: int) -> Optional[ContractDetail]:
        """Get a single contract by ID with detailed information."""
        contract = self.db.query(Contract).filter(Contract.id == contract_id).first()
        if not contract:
            return None

        risk = None
        if contract.risk_assessment:
            risk = RiskOut(
                crs=contract.risk_assessment.crs,
                risk_level=self._get_risk_level(contract),
                rule_score=contract.risk_assessment.rule_score,
                anomaly_score=contract.risk_assessment.anomaly_score,
                flags=[
                    RiskFlagOut(
                        flag_id=f.flag_id,
                        detected=f.detected,
                        severity=f.severity,
                        score=f.score,
                        explanation=f.explanation
                    ) for f in contract.risk_flags if f.detected
                ]
            )

        return ContractDetail(
            id=contract.id,
            contract_number=contract.contract_number,
            title=contract.title,
            contract_date=contract.contract_date,
            department_id=contract.department_id,
            vendor_id=contract.vendor_id,
            estimate_value=contract.estimate_value,
            award_value=contract.award_value,
            specification=contract.specification,
            tender_start=contract.tender_start,
            tender_end=contract.tender_end,
            bidder_count=len(contract.bids),
            risk=risk
        )

    def analyze_contract_risk(self, contract_id: int) -> dict:
        """Perform risk analysis on a contract and store results."""
        contract = self.db.query(Contract).filter(Contract.id == contract_id).first()
        if not contract:
            raise ValueError(f"Contract with ID {contract_id} not found")

        return self.risk_engine.analyze_contract(contract, self.db)

    def _get_risk_level(self, contract: Contract) -> str:
        """Get risk level based on CRS score."""
        if not contract.risk_assessment:
            return "unknown"

        crs = contract.risk_assessment.crs
        if crs >= settings.risk_threshold:
            return "high"
        elif crs >= 40:
            return "medium"
        else:
            return "low"

File: src/services/contractService.js
import { api } from "./api";

/**
 * @typedef {Object} ContractQueryParams
 * @property {number} [department_id] - Filter by procuring department ID
 * @property {number} [vendor_id] - Filter by awarded vendor ID
 * @property {"high"|"medium"|"low"} [risk_level] - Filter by risk severity band
 * @property {string} [search] - Full-text search by contract number or title
 * @property {number} [limit=50] - Number of records to return
 * @property {number} [offset=0] - Pagination offset
 */

/**
 * @typedef {Object} RiskFlag
 * @property {string} flag_id - Risk indicator code (e.g. RF-1)
 * @property {boolean} detected - Whether this flag was triggered
 * @property {string} severity - Severity level: "high", "medium", "low"
 * @property {number} score - Contribution score
 * @property {string} explanation - Plain-language explanation
 */

/**
 * @typedef {Object} PeerComparison
 * @property {number} department_total_contracts
 * @property {number} peer_median_award_value
 * @property {number} peer_mean_award_value
 * @property {number} value_deviation_percent
 * @property {number} peer_median_tender_days
 * @property {number} duration_deviation_percent
 * @property {number} peer_average_bidders
 * @property {boolean} is_value_outlier
 * @property {boolean} is_duration_outlier
 * @property {string} explanation
 */

/**
 * @typedef {Object} ContractSummary
 * @property {number} id
 * @property {string} contract_number
 * @property {string} title
 * @property {string} contract_date
 * @property {number} department_id
 * @property {string|null} department_name
 * @property {number} vendor_id
 * @property {string|null} vendor_name
 * @property {number} estimate_value
 * @property {number} award_value
 * @property {number|null} crs
 * @property {"high"|"medium"|"low"|null} risk_level
 */

/**
 * @typedef {Object} SimilarTender
 * @property {number} contract_id
 * @property {string} contract_number
 * @property {string} title
 * @property {string} department_name
 * @property {string} vendor_name
 * @property {number} award_value
 * @property {number} similarity_score
 * @property {string[]} matched_terms
 */

/**
 * @typedef {Object} RuleEvidence
 * @property {string} rule_id
 * @property {string} rule_name
 * @property {boolean} triggered
 * @property {string} severity
 * @property {number|null} contribution
 * @property {string} explanation
 * @property {Object|null} evidence
 */

/**
 * @typedef {Object} RiskEvidence
 * @property {number} contract_id
 * @property {number} risk_score
 * @property {string} risk_level
 * @property {number} rule_score
 * @property {number} anomaly_score
 * @property {RuleEvidence[]} triggered_rules
 * @property {string} generated_at
 */

export const contractService = {
  /**
   * List contracts with optional search, risk filter, and pagination.
   * @param {ContractQueryParams} [params={}]
   * @returns {Promise<ContractSummary[]>}
   */
  async getContracts(params = {}) {
    const response = await api.get("/contracts", { params });
    return response.data;
  },

  /**
   * Get complete contract dossier with bids, risk flags, extensions, and peer comparison.
   * @param {number|string} contractId
   * @returns {Promise<Object>}
   */
  async getContract(contractId) {
    const response = await api.get(`/contracts/${contractId}`);
    return response.data;
  },

  /**
   * Search for recycled or verbatim tender specifications using TF-IDF cosine similarity.
   * @param {number|string} contractId
   * @param {number} [limit=5]
   * @returns {Promise<SimilarTender[]>}
   */
  async getSimilarTenders(contractId, limit = 5) {
    const response = await api.get(`/contracts/${contractId}/similar-tenders`, {
      params: { limit },
    });
    return response.data;
  },

  /**
   * Retrieve granular forensic risk evidence and red-flag explanations.
   * @param {number|string} contractId
   * @returns {Promise<RiskEvidence>}
   */
  async getRiskEvidence(contractId) {
    const response = await api.get(`/contracts/${contractId}/risk-evidence`);
    return response.data;
  },
};

export default contractService;
